python/pylibvw.cc

```diff
@@ -655,39 +711,67 @@ void ex_push_dictionary(example_ptr ec, vw_ptr vw, py::dict& dict)
     if (ns_e().length() < 1) continue;
     py::extract<py::list> list_e(objectVal);
     py::list list = list_e();
-    char ns = ns_e()[0];
+    std::string ns_str = ns_e();
+    char* ns = new char [ns_str.length()+1];
+    std::strcpy (ns, ns_str.c_str());
     ex_ensure_namespace_exists(ec, ns);
     ex_push_feature_list(ec, vw, ns, list);
   }
 }
 
-bool ex_pop_feature(example_ptr ec, unsigned char ns)
+bool ex_pop_feature(example_ptr ec, char* ns)
 {
-  if (ec->feature_space[ns].size() == 0) return false;
-  float val = ec->feature_space[ns].values.back();
-  ec->feature_space[ns].values.pop_back();
-  if (ec->feature_space[ns].indicies.size() > 0) ec->feature_space[ns].indicies.pop_back();
-  if (ec->feature_space[ns].space_names.size() > 0) ec->feature_space[ns].space_names.pop_back();
+  uint64_t hash = hashstring(ns, strlen(ns), 0);
+  features& feats = ec->feature_space[hash];
+  if (feats.size() == 0) return false;
+  float val = feats.values.back();
+  feats.values.pop_back();
+  if (feats.indicies.size() > 0) feats.indicies.pop_back();
+  if (feats.space_names.size() > 0) feats.space_names.pop_back();
   ec->num_features--;
-  ec->feature_space[ns].sum_feat_sq -= val * val;
+  feats.sum_feat_sq -= val * val;
   ec->reset_total_sum_feat_sq();
   return true;
 }
 
-void ex_erase_namespace(example_ptr ec, unsigned char ns)
+void ex_erase_namespace(example_ptr ec, char* ns)
 {
-  ec->num_features -= ec->feature_space[ns].size();
+  if (strlen(ns) == 1)
+  {
+    auto beg = ec->feature_space.namespace_index_begin(ns[0]);
+    auto end = ec->feature_space.namespace_index_end(ns[0]);
+    std::vector<uint64_t> rem_hashes;
+    for (auto it = beg; it != end; ++it) {
+      features& feats = *it;
+      ec->num_features -= feats.size();
+      feats.sum_feat_sq = 0.f;
+      feats.clear();
```

Comment:
There's no need to do the sum_feat_sq and clear operations if you're about to remove it